### DISCUSSION TITLE
[Bug] hidden_channels=c["hidden_channels"], KeyError: 'hidden_channels'

### DISCUSSION THREAD

**Discussion by zubairahmed-ai** [coqui-ai/TTS#536] [General Q&A]
**Describe the bug**
When I am trying to train `speedy_speech` on LJSpeech dataset like this `~/work/TTS-0.0.13$ CUDA_VISIBLE_DEVICES="0" python TTS/bin/train_speedy_speech.py --config_path config.json | tee training.log` I get the following error

```
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
Traceback (most recent call last):
  File "TTS/bin/train_speedy_speech.py", line 570, in <module>
    main(args)
  File "TTS/bin/train_speedy_speech.py", line 481, in main
    model = setup_model(num_chars, num_speakers, c, speaker_embedding_dim=speaker_embedding_dim)
  File "/home/ubuntu/work/TTS-0.0.13/TTS/tts/utils/generic_utils.py", line 140, in setup_model
    hidden_channels=c["hidden_channels"],
KeyError: 'hidden_channels'
 > Using CUDA:  True
 > Number of GPUs:  1
   >  Mixed precision mode is ON
 > Git Hash: 0000000
 > Experiment folder: /home/ubuntu/work/TTS-0.0.13/output/ljspeech-speedy-speech-June-04-2021_10+42AM-0000000
 > Setting up Audio Processor...
 | > sample_rate:22050
 | > resample:False
 | > num_mels:80
 | > min_level_db:-100
 | > frame_shift_ms:None
 | > frame_length_ms:None
 | > ref_level_db:20
 | > fft_size:1024
 | > power:1.5
 | > preemphasis:0.0
 | > griffin_lim_iters:60
 | > signal_norm:True
 | > symmetric_norm:True
 | > mel_fmin:50.0
 | > mel_fmax:7600.0
 | > spec_gain:1.0
 | > stft_pad_mode:reflect
 | > max_norm:4.0
 | > clip_norm:True
 | > do_trim_silence:True
 | > trim_db:60
 | > do_sound_norm:False
 | > stats_path:None
 | > base:10
 | > hop_length:256
 | > win_length:1024
 | > Found 13100 files in /home/ubuntu/work/TTS-0.0.13/LJSpeech-1.1
 > Using model: speedy_speech
 ! Run is removed from /home/ubuntu/work/TTS-0.0.13/output/ljspeech-speedy-speech-June-04-2021_10+42AM-0000000

```

**To Reproduce**
1. Run the following command `CUDA_VISIBLE_DEVICES="0" python TTS/bin/train_speedy_speech.py --config_path config.json | tee training.log`

**Expected behavior**
Must start training

**Environment (please complete the following information):**
- **Ubuntu 18.04**:
- **TensorFlow**:
- **Python 3.6**:

**Comment by zubairahmed-ai** [comment]
[training.log](https://github.com/coqui-ai/TTS/files/6597917/training.log)

[config.json.txt](https://github.com/coqui-ai/TTS/files/6597918/config.json.txt)

Here's the `training.log` and `config.json`

**Comment by lexkoro** [comment]
Your config seems like its for tacotron, which doesn't have the parameter `hidden_channels`.
See error message `KeyError: 'hidden_channels'`
 
There should also be an example config for speedyspeech.

**Comment by zubairahmed-ai** [comment]
It is for tacotron, where to find the example config for speedy speech?

**Comment by zubairahmed-ai** [comment]
Found the config. trying now

**Comment by zubairahmed-ai** [comment]
deleted my earlier comment
I am getting the missing wav file error, I downloaded the dataset and there are over 13k files in it but not this one

```
Traceback (most recent call last):
  File "TTS/bin/train_speedy_speech.py", line 570, in <module>
    main(args)
  File "TTS/bin/train_speedy_speech.py", line 469, in main
    meta_data_train, meta_data_eval = load_meta_data(c.datasets, eval_split=True)
  File "/home/ubuntu/work/TTS-0.0.13/TTS/tts/datasets/preprocess.py", line 43, in load_meta_data
    attn_file = meta_data[ins[1]].strip()
KeyError: 'LJSpeech-1.1/wavs/LJ023-0005.wav'
 > Using CUDA:  True
 > Number of GPUs:  1
 > Git Hash: 0000000
 > Experiment folder: /home/ubuntu/work/TTS-0.0.13/output/speedy-speech-ljspeech-June-04-2021_02+15PM-0000000
 > Setting up Audio Processor...
 | > sample_rate:22050
 | > resample:False
 | > num_mels:80
 | > min_level_db:-100
 | > frame_shift_ms:None
 | > frame_length_ms:None
 | > ref_level_db:20
 | > fft_size:1024
 | > power:1.5
 | > preemphasis:0.0
 | > griffin_lim_iters:60
 | > signal_norm:True
 | > symmetric_norm:True
 | > mel_fmin:50.0
 | > mel_fmax:7600.0
 | > spec_gain:1.0
 | > stft_pad_mode:reflect
 | > max_norm:4.0
 | > clip_norm:True
 | > do_trim_silence:True
 | > trim_db:60
 | > do_sound_norm:False
 | > stats_path:None
 | > base:10
 | > hop_length:256
 | > win_length:1024
 | > Found 13100 files in /home/ubuntu/work/TTS-0.0.13/LJSpeech-1.1
 ! Run is removed from /home/ubuntu/work/TTS-0.0.13/output/speedy-speech-ljspeech-June-04-2021_02+15PM-0000000

```

**Reply by Sadam1195** [reply]
![Screenshot from 2021-06-12 12-59-01](https://user-images.githubusercontent.com/42005385/121769613-0713fb80-cb7e-11eb-8a1a-9314ac136b7f.png)

This file is in the dataset. I was getting similar error on my dataset simple fix was copying metadata.csv file inside wavs folder. Do try that.
@zubairahmed-ai

**Reply by zubairahmed-ai** [reply]
Thanks so much @Sadam1195 for replying, I have actually moved to trying AlignTTS as @erogol suggested below, I haven't started training that yet and I am sure some issues await me, have you trained AlignTTS though and faced any issues?

**Reply by Sadam1195** [reply]
> Thanks so much @Sadam1195 for replying, I have actually moved to trying AlignTTS as @erogol suggested below, I haven't started training that yet and I am sure some issues await me, have you trained AlignTTS though and faced any issues?

No, I haven't trained AllignTTS yet as currently I have two models in training....

**Reply by zubairahmed-ai** [reply]
Thanks so much for quickly replying to me
Since I am a beginner could you please help me understand by answering the below newbie questions I have
How big is your dataset?
How long you roughly took to record it?
How long you train before you realistically expect good performance?
What sort of compute power/GPU do you train with?
Which model are you training?

**Reply by Sadam1195** [reply]
> Thanks so much for quickly replying to me Since I am a beginner could you please help me understand by answering the below newbie questions I have

How big is your dataset?
>31h 45m

How long you roughly took to record it?
>It's a public dataset so I don't know how much time it took them to record it. Probably days or few weeks as often recording is done in sessions. 

How long you train before you realistically expect good performance?
>That's relative and depends on which model you are training. 
- With Tacotron2 you can expect to hear audible speech after 20k steps but 150k-300k should be minimum.
- With WaveRNN you need to at least train the model for 220k for good quality.

What sort of compute power/GPU do you train with?
>I have this GPU which has 16 GB RAM.

![GPU](https://user-images.githubusercontent.com/42005385/121771699-62e48180-cb8a-11eb-8d8c-155e9cd0f6c1.png)

Which model are you training?
>I am training Tacotron2 and WaveRNN.  (Tacotron2 and WaveRNN are the best with the results they produce (in my experience) but they are a lot more time consuming to train and specially WaveRNN is very slow when inference but its quality is unmatched and Tacotron2 is hard to train as often it doesn't converge with default config settings.)

**Reply by zubairahmed-ai** [reply]
Thanks @Sadam1195 for sharing this

**Comment by zubairahmed-ai** [comment]
Note that it prompted for a path of `metadata_attn_mask.txt` so I gave it to `tests/data/ljspeech/metadata_attn_mask.txt` which I suppose is wrong so not sure what to put there?

**Reply by lexkoro** [reply]
I moved the post to discussions, since it is not really a bug/ issue but rather a question.

**Comment by lexkoro** [comment]
Sorry I haven't trained speedy speech myself, so not quite sure what is needed here. 

But from my understanding, you can compute the attention mask with this script here:
https://github.com/coqui-ai/TTS/blob/main/TTS/bin/compute_attention_masks.py

But this will require a functional tacotron model first I guess?

**Reply by zubairahmed-ai** [reply]
Yea this much I figured out too but not sure exactly what's going on here☹️

**Reply by zubairahmed-ai** [reply]
Does @erogol @Edresson can you please help with this? I am unable to train speedy_speech model?

**Reply by erogol** [reply]
you need to use compute attention masks with a functional Tacotron1-2 model by `compute_attention_masks.py`. It produces masks and a meta file. You need to give this metafile to 'meta_file_attn_mask' field in config. Then things should work. 

It is not optimal but you can check speedy speech related tests to see how it works if you need more info.

**Reply by zubairahmed-ai** [reply]
Thanks @erogol so I can use any pretrained Tacotron 2 model or does it need to be trained with my own data first before training Speedy speech?

**Reply by erogol** [reply]
yes you need to train on you data first. Or use AlignTTS which is similar to SS but it aligns itself.

**Reply by zubairahmed-ai** [reply]
Darn that's hard, I will check this

**Reply by zubairahmed-ai** [reply]
@erogol which one of these models is Align-TTS?

```
Name format: type/language/dataset/model
 >: tts_models/en/ek1/tacotron2 [already downloaded]
 >: tts_models/en/ljspeech/tacotron2-DDC [already downloaded]
 >: tts_models/en/ljspeech/glow-tts [already downloaded]
 >: tts_models/en/ljspeech/tacotron2-DCA [already downloaded]
 >: tts_models/en/ljspeech/speedy-speech-wn [already downloaded]
 >: tts_models/en/vctk/sc-glow-tts [already downloaded]
 >: tts_models/en/sam/tacotron-DDC
 >: tts_models/es/mai/tacotron2-DDC
 >: tts_models/fr/mai/tacotron2-DDC
 >: tts_models/zh-CN/baker/tacotron2-DDC-GST
 >: tts_models/nl/mai/tacotron2-DDC
 >: tts_models/ru/ruslan/tacotron2-DDC
 >: tts_models/de/thorsten/tacotron2-DCA
 >: vocoder_models/universal/libri-tts/wavegrad [already downloaded]
 >: vocoder_models/universal/libri-tts/fullband-melgan [already downloaded]
 >: vocoder_models/en/ek1/wavegrad [already downloaded]
 >: vocoder_models/en/ljspeech/multiband-melgan [already downloaded]
 >: vocoder_models/en/ljspeech/hifigan_v2 [already downloaded]
 >: vocoder_models/en/vctk/hifigan_v2 [already downloaded]
 >: vocoder_models/en/sam/hifigan_v2 [already downloaded]
 >: vocoder_models/nl/mai/parallel-wavegan
 >: vocoder_models/de/thorsten/wavegrad

```

**Reply by zubairahmed-ai** [reply]
ok that's `glow-tts`

**Reply by zubairahmed-ai** [reply]
I am trying to train `align_tts`, I coped the config from `glow_tts_gated_conv.json` to `aligntts_transformers.json` because the file was empty but I am still getting the below error

```
~/work/TTS-0.0.13$ CUDA_VISIBLE_DEVICES="0" python TTS/bin/train_align_tts.py --config_path config.json | tee training.log
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
Traceback (most recent call last):
  File "TTS/bin/train_align_tts.py", line 562, in <module>
    main(args)
  File "TTS/bin/train_align_tts.py", line 458, in main
    model = setup_model(num_chars, num_speakers, c, speaker_embedding_dim=speaker_embedding_dim)
  File "/home/ubuntu/work/TTS-0.0.13/TTS/tts/utils/generic_utils.py", line 152, in setup_model
    hidden_channels=c["hidden_channels"],
KeyError: 'hidden_channels'
 > Using CUDA:  True
 > Number of GPUs:  1
 > Git Hash: 0000000
 > Experiment folder: /home/ubuntu/work/TTS-0.0.13/output/align_tts-June-07-2021_07+40AM-0000000
 > Setting up Audio Processor...
 | > sample_rate:22050
 | > resample:False
 | > num_mels:80
 | > min_level_db:-100
 | > frame_shift_ms:None
 | > frame_length_ms:None
 | > ref_level_db:0
 | > fft_size:1024
 | > power:1.1
 | > preemphasis:0.0
 | > griffin_lim_iters:60
 | > signal_norm:True
 | > symmetric_norm:True
 | > mel_fmin:50.0
 | > mel_fmax:7600.0
 | > spec_gain:1.0
 | > stft_pad_mode:reflect
 | > max_norm:1.0
 | > clip_norm:True
 | > do_trim_silence:True
 | > trim_db:60
 | > do_sound_norm:False
 | > stats_path:None
 | > base:10
 | > hop_length:256
 | > win_length:1024
 | > Found 44063 files in /home/ubuntu/work/TTS-0.0.13/VCTK-Corpus-removed-silence
 > Using model: align_tts
 ! Run is removed from /home/ubuntu/work/TTS-0.0.13/output/align_tts-June-07-2021_07+40AM-0000000

```

Here's my [config.json](https://github.com/coqui-ai/TTS/files/6606836/config.json.txt)

**Reply by erogol** [reply]
check this file and fill the fields in your `config.json` according to the docstring comments. 
https://github.com/coqui-ai/TTS/blob/main/TTS/tts/configs/align_tts_config.py

**Reply by zubairahmed-ai** [reply]
Thanks @erogol 
Training has started.  [training.log](https://github.com/coqui-ai/TTS/files/6608101/training.log.txt)

I misunderstood your above comment and didn't know you meant to add all values above in my own config file before training begins, I included that and even ran into [this resample param issue](https://github.com/coqui-ai/TTS/issues/483#issuecomment-840850964)

Here's my config for anyone else who is interested to train

[aligntts_transformers.json](https://github.com/coqui-ai/TTS/files/6608066/aligntts_transformers.json.txt)
[config.json](https://github.com/coqui-ai/TTS/files/6608072/config.json.txt)

After training this I will try the speaker encoder that @Edresson [recommends here](https://github.com/coqui-ai/TTS/issues/483#issuecomment-842267443)

**Reply by iprovalo** [reply]
@erogol is the requirement the same for the Fast Speech model?

> you need to use compute attention masks with a functional Tacotron1-2 model by `compute_attention_masks.py`. It produces masks and a meta file. You need to give this metafile to 'meta_file_attn_mask' field in config. Then things should work.

**Comment by zubairahmed-ai** [comment]
@erogol what could cause this error? I didn't change anything else except config 
@The0nix did you face this?
```
> EVALUATION

Traceback (most recent call last):
  File "TTS/bin/train_align_tts.py", line 562, in <module>
    main(args)
  File "TTS/bin/train_align_tts.py", line 539, in main
    eval_avg_loss_dict = evaluate(eval_loader, model, criterion, ap, global_step, epoch, cur_phase)
  File "/home/ubuntu/anaconda3/envs/ttstf/lib/python3.6/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "TTS/bin/train_align_tts.py", line 349, in evaluate
    pred_spec = decoder_output[idx].detach().data.cpu().numpy().T
TypeError: 'NoneType' object is not subscriptable
 ! Run is removed from /home/ubuntu/work/TTS-0.0.13/output/align_tts-June-07-2021_04+14PM-0000000
```

**Reply by zubairahmed-ai** [reply]
> @erogol what could cause this error? I didn't change anything else except config
> @The0nix did you face this?
> 
> ```
> > EVALUATION
> 
> Traceback (most recent call last):
>   File "TTS/bin/train_align_tts.py", line 562, in <module>
>     main(args)
>   File "TTS/bin/train_align_tts.py", line 539, in main
>     eval_avg_loss_dict = evaluate(eval_loader, model, criterion, ap, global_step, epoch, cur_phase)
>   File "/home/ubuntu/anaconda3/envs/ttstf/lib/python3.6/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
>     return func(*args, **kwargs)
>   File "TTS/bin/train_align_tts.py", line 349, in evaluate
>     pred_spec = decoder_output[idx].detach().data.cpu().numpy().T
> TypeError: 'NoneType' object is not subscriptable
>  ! Run is removed from /home/ubuntu/work/TTS-0.0.13/output/align_tts-June-07-2021_04+14PM-0000000
> ```

For anyone reading this the error was gone after doing [this](https://github.com/coqui-ai/TTS/issues/556#issuecomment-856722786)

**Comment by zubairahmed-ai** [comment]
@erogol I have a small dataset now, I like to do Transfer learning by fine-tuning AlignTTS, where can I download a pretrained model on VCTK dataset and the path should I put in the `--restore-path` below?

`python TTS/bin/train_align_tts.py --config_path TTS/tts/configs/config.json --restore_path /path/to/your/model.pth.tar`

**Reply by thorstenMueller** [reply]
You can check the released models in Coqui TTS wiki.
https://github.com/coqui-ai/TTS/wiki/Experimental-Released-Models#english

Checkpoint and config files download (_Google drive_) for VCTK dataset are linked there.

**Reply by zubairahmed-ai** [reply]
@thorstenMueller I am confused, what about these https://github.com/coqui-ai/TTS/blob/main/TTS/.models.json it also seems to have Align-TTS/Sc-Glow-TTS trained on VCTK dataset?
@Edresson any help?

**Reply by erogol** [reply]
Experimental models do not promise compatibility but the ones in `.models.json` do

**Reply by zubairahmed-ai** [reply]
I just started training from the models.json one of AlignTTS
Which one to use to fine-tune on my dataset?
